File: dart/worldbank-gdp.dart
// To parse this JSON data, do
//
//     final gdp = gdpFromJson(jsonString);

import 'dart:convert';

List<dynamic> gdpFromJson(String str) {
    final jsonData = json.decode(str);
    return new List<dynamic>.from(jsonData.map((x) => x));
}

String gdpToJson(List<dynamic> data) {
    final dyn = new List<dynamic>.from(data.map((x) => x));
    return json.encode(dyn);
}

class GdpElement {
    Country indicator;
    Country country;
    String value;
    String decimal;
    String date;

    GdpElement({
        this.indicator,
        this.country,
        this.value,
        this.decimal,
        this.date,
    });

    factory GdpElement.fromJson(Map<String, dynamic> json) => new GdpElement(
        indicator: Country.fromJson(json["indicator"]),
        country: Country.fromJson(json["country"]),
        value: json["value"],
        decimal: json["decimal"],
        date: json["date"],
    );

    Map<String, dynamic> toJson() => {
        "indicator": indicator.toJson(),
        "country": country.toJson(),
        "value": value,
        "decimal": decimal,
        "date": date,
    };
}

class Country {
    Id id;
    Value value;

    Country({
        this.id,
        this.value,
    });

    factory Country.fromJson(Map<String, dynamic> json) => new Country(
        id: idValues.map[json["id"]],
        value: valueValues.map[json["value"]],
    );

    Map<String, dynamic> toJson() => {
        "id": idValues.reverse[id],
        "value": valueValues.reverse[value],
    };
}

enum Id { CN, US, NY_GDP_MKTP_CD }

final idValues = new EnumValues({
    "CN": Id.CN,
    "NY.GDP.MKTP.CD": Id.NY_GDP_MKTP_CD,
    "US": Id.US
});

enum Value { CHINA, UNITED_STATES, GDP_CURRENT_US }

final valueValues = new EnumValues({
    "China": Value.CHINA,
    "GDP (current US\u0024)": Value.GDP_CURRENT_US,
    "United States": Value.UNITED_STATES
});

class PurpleGdp {
    int page;
    int pages;
    String perPage;
    int total;

    PurpleGdp({
        this.page,
        this.pages,
        this.perPage,
        this.total,
    });

    factory PurpleGdp.fromJson(Map<String, dynamic> json) => new PurpleGdp(
        page: json["page"],
        pages: json["pages"],
        perPage: json["per_page"],
        total: json["total"],
    );

    Map<String, dynamic> toJson() => {
        "page": page,
        "pages": pages,
        "per_page": perPage,
        "total": total,
    };
}

class EnumValues<T> {
    Map<String, T> map;
    Map<T, String> reverseMap;

    EnumValues(this.map);

    Map<T, String> get reverse {
        if (reverseMap == null) {
            reverseMap = map.map((k, v) => new MapEntry(v, k));
        }
        return reverseMap;
    }
}

File: kotlin/worldbank-gdp.kt
// To parse the JSON, install Klaxon and do:
//
//   val gdp = Gdp.fromJson(jsonString)

package quicktype

import com.beust.klaxon.*

private fun <T> Klaxon.convert(k: kotlin.reflect.KClass<*>, fromJson: (JsonValue) -> T, toJson: (T) -> String, isUnion: Boolean = false) =
    this.converter(object: Converter {
        @Suppress("UNCHECKED_CAST")
        override fun toJson(value: Any)        = toJson(value as T)
        override fun fromJson(jv: JsonValue)   = fromJson(jv) as Any
        override fun canConvert(cls: Class<*>) = cls == k.java || (isUnion && cls.superclass == k.java)
    })

private val klaxon = Klaxon()
    .convert(ID::class,       { ID.fromValue(it.string!!) },    { "\"${it.value}\"" })
    .convert(Value::class,    { Value.fromValue(it.string!!) }, { "\"${it.value}\"" })
    .convert(GDPUnion::class, { GDPUnion.fromJson(it) },        { it.toJson() }, true)

class Gdp(elements: Collection<GDPUnion>) : ArrayList<GDPUnion>(elements) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = Gdp(klaxon.parseArray<GDPUnion>(json)!!)
    }
}

sealed class GDPUnion {
    class FluffyGDPValue(val value: FluffyGDP)            : GDPUnion()
    class PurpleGDPArrayValue(val value: List<PurpleGDP>) : GDPUnion()

    public fun toJson(): String = klaxon.toJsonString(when (this) {
        is FluffyGDPValue      -> this.value
        is PurpleGDPArrayValue -> this.value
    })

    companion object {
        public fun fromJson(jv: JsonValue): GDPUnion = when (jv.inside) {
            is JsonObject   -> FluffyGDPValue(jv.obj?.let { klaxon.parseFromJsonObject<FluffyGDP>(it) }!!)
            is JsonArray<*> -> PurpleGDPArrayValue(jv.array?.let { klaxon.parseFromJsonArray<PurpleGDP>(it) }!!)
            else            -> throw IllegalArgumentException()
        }
    }
}

data class PurpleGDP (
    val indicator: Country,
    val country: Country,
    val value: String,
    val decimal: String,
    val date: String
)

data class Country (
    val id: ID,
    val value: Value
)

enum class ID(val value: String) {
    CN("CN"),
    NyGdpMktpCD("NY.GDP.MKTP.CD"),
    Us("US");

    companion object {
        public fun fromValue(value: String): ID = when (value) {
            "CN"             -> CN
            "NY.GDP.MKTP.CD" -> NyGdpMktpCD
            "US"             -> Us
            else             -> throw IllegalArgumentException()
        }
    }
}

enum class Value(val value: String) {
    China("China"),
    GDPCurrentUS("GDP (current US\$)"),
    UnitedStates("United States");

    companion object {
        public fun fromValue(value: String): Value = when (value) {
            "China"              -> China
            "GDP (current US\$)" -> GDPCurrentUS
            "United States"      -> UnitedStates
            else                 -> throw IllegalArgumentException()
        }
    }
}

data class FluffyGDP (
    val page: Long,
    val pages: Long,

    @Json(name = "per_page")
    val perPage: String,

    val total: Long
)
